app/src/main/java/protect/budgetwatch/PasswordManager.java

```diff
@@ -0,0 +1,98 @@
+package protect.budgetwatch;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import java.math.BigInteger;
+import java.security.SecureRandom;
+
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+
+public class PasswordManager {
+
+    private final String PasswordKey = "password";
+    private final int Iterations = 100;
+    private final int KeyLength = 128;
+
+    private SharedPreferences _prefs;
+
+    PasswordManager(Context context) {
+        _prefs = context.getSharedPreferences("protect.budgetwatch", Context.MODE_PRIVATE);
+    }
+
+    private String getStoredPasswordHash() {
+        return _prefs.getString(PasswordKey, null);
+    }
+
+    public void setPassword(String password) {
+        try {
+            // Generate Salt
+            SecureRandom sr = SecureRandom.getInstance("SHA1PRNG");
+            byte[] salt = new byte[KeyLength / 8];
+            sr.nextBytes(salt);
+            // Generate Hash
+            byte[] hash = generateHash(password, salt);
+            // Write to SharedPreferences
+            _prefs.edit().putString(PasswordKey, toHex(salt) + ":" + toHex(hash)).commit();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    void clearPassword() {
+        _prefs.edit().remove(PasswordKey).commit();
+    }
+
+    boolean isPasswordEnabled() {
+        return getStoredPasswordHash() != null;
+    }
+
+    boolean isPasswordCorrect(String password) {
+        String storedPassword = getStoredPasswordHash();
+        if (storedPassword == null) return false;
+
+        String[] parts = storedPassword.split(":");
+        byte[] salt = fromHex(parts[0]);
+        byte[] hash = fromHex(parts[1]);
+
+        byte[] passwordHash = generateHash(password, salt);
+
+        int diff = hash.length ^ passwordHash.length;
+        for (int i = 0; i < hash.length && i < passwordHash.length; i++) {
+            diff |= hash[i] ^ passwordHash[i];
+        }
+        return diff == 0;
+    }
+
+    private byte[] generateHash(String password, byte[] salt) {
+        PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, Iterations, KeyLength);
+        try {
+            SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
+            return skf.generateSecret(spec).getEncoded();
+        } catch (Exception e) {
+            e.printStackTrace();
+            return new byte[0];
```

Comment:
I'm thinking if generating the key failed the error should be presented to the user, instead of attempting to save an empty password. I simulated this failing by editing the preferences file and setting an empty password hash. The result was when a password was entered to authorize when it was compared an out of bounds exception resulted:
```
01-21 23:43:40.888  6693  6693 E AndroidRuntime: FATAL EXCEPTION: main
01-21 23:43:40.888  6693  6693 E AndroidRuntime: Process: protect.budgetwatch, PID: 6693
01-21 23:43:40.888  6693  6693 E AndroidRuntime: java.lang.ArrayIndexOutOfBoundsException: length=1; index=1
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at protect.budgetwatch.PasswordManager.isPasswordCorrect(PasswordManager.java:58)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at protect.budgetwatch.PasswordAuthenticationActivity.attemptLogin(PasswordAuthenticationActivity.java:65)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at protect.budgetwatch.PasswordAuthenticationActivity.access$000(PasswordAuthenticationActivity.java:14)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at protect.budgetwatch.PasswordAuthenticationActivity$2.onClick(PasswordAuthenticationActivity.java:44)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at android.view.View.performClick(View.java:6597)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at android.view.View.performClickInternal(View.java:6574)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at android.view.View.access$3100(View.java:778)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at android.view.View$PerformClick.run(View.java:25885)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:873)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:193)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6669)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
01-21 23:43:40.888  6693  6693 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```